packages/jupyterlab-manager/src/semverregistry.ts

```diff
@@ -0,0 +1,147 @@
+    if (versions === undefined) {
+      versions = new Map<string, T>();
+      this._registry.set(name, versions);
+    }
+
+    if (versions.has(version)) {
+      throw new Error(`Version ${version} of ${name} already registered.`);
+    }
+    versions.set(version, object);
+
+    // Some pending requests may now be satisfied, so schedule a retry.
+    if (this._pendingRequests.length > 0) {
+      this._throttle.invoke();
+    }
+  }
+
+  /**
+   * Get an object for a name and semVer range.
+   *
+   * @returns a promise resolving to the object with the maximum version
+   * satisfying the version range if one is registered, or undefined if one is
+   * not registered.
+   *
+   * #### Notes
+   * Resolution may be delayed for a short timeout to allow for some delay in
+   * registering the object.
+   */
+  async get(name: string, semVer: string): Promise<T | undefined> {
+    const mod = this._getSync(name, semVer);
+    if (mod) {
+      return mod;
+    }
+    const result = new PromiseDelegate<T | undefined>();
+    const timestamp = Date.now();
+    this._pendingRequests.push({ name, semVer, result, timestamp });
+    this._throttle.invoke();
+    return result.promise;
+  }
+
+  /**
+   * Synchronously get an object for a name and semVer range.
+   *
+   * @returns the object with the maximum version satisfying the version range
+   * or undefined if one does not exist.
+   */
+  private _getSync(name: string, semVer: string): T | undefined {
+    const versions = this._registry.get(name);
+    if (versions) {
+      const best = maxSatisfying(Array.from(versions.keys()), semVer);
+      if (best !== null) {
+        return versions.get(best);
+      }
+    }
+  }
+
+  /**
+   * Retry resolving pending requests.
+   */
+  private _retry(): void {
+    const now = Date.now();
+    // Try to resolve each pending request successfully.
+    this._pendingRequests.forEach(request => {
+      const { name: key, semVer: semver, result } = request;
+      const mod = this._getSync(key, semver);
+      if (mod) {
+        result.resolve(mod);
+        request.timestamp = null;
+      }
+    });
+
+    // Clear out all requests which were fulfilled or expired.
+    ArrayExt.removeAllWhere(
+      this._pendingRequests,
+      ({ result, timestamp }) => {
+        if (timestamp === null) {
+          // Request fulfilled, so remove the entry.
+          return true;
+        } else if (now - timestamp > TIMEOUT) {
+          // Request expired, so give up and remove the entry.
+          result.resolve(undefined);
+          return true;
+        }
+        return false;
+      }
+    );
+
+    // If we still have pending requests, try again later.
+    if (this._pendingRequests.length > 0) {
+      this._throttle.invoke();
+    }
+  }
+
+  private _registry = new Map<string, Map<string, T>>();
+  private _pendingRequests: IRequest<T>[] = [];
+  private _throttle = new Throttler(() => {
+    this._retry();
+  }, TIMEOUT);
```

Comment:
I really prefer initializers with actual logic to be specified in the constructor 😕 (I prefer most initializers to be in the constructor actually, but when there is good typing they are less important).

Reply:
Noted. I prefer what can be moved to the definition to be there.